src/fu-remotes.c

```diff
@@ -0,0 +1,503 @@
+			if (!xb_builder_source_load_file (source, file,
+							  XB_BUILDER_SOURCE_FLAG_NONE,
+							  NULL, error))
+				return FALSE;
+			xb_builder_import_source (builder, source);
+		}
+	}
+	return TRUE;
+}
+
+gboolean
+fu_remotes_load (FuRemotes *self, FuRemotesLoadFlags flags, GError **error)
+{
+	const gchar *const *locales = g_get_language_names ();
+	g_autofree gchar *cachedirpkg = NULL;
+	g_autofree gchar *xmlbfn = NULL;
+	g_autoptr(GFile) xmlb = NULL;
+	g_autoptr(XbBuilder) builder = xb_builder_new ();
+	XbBuilderCompileFlags compile_flags = XB_BUILDER_COMPILE_FLAG_SINGLE_LANG |
+					      XB_BUILDER_COMPILE_FLAG_IGNORE_INVALID;
+
+	g_return_val_if_fail (FU_IS_REMOTES (self), FALSE);
+	g_return_val_if_fail (self->os_release == NULL, FALSE);
+
+	/* load AppStream about the remotes */
+	self->os_release = fwupd_get_os_release (error);
+	if (self->os_release == NULL)
+		return FALSE;
+	if (!fu_remotes_load_metainfos (builder, error))
+		return FALSE;
+
+	/* add the locales, which is really only going to be 'C' or 'en' */
+	for (guint i = 0; locales[i] != NULL; i++)
+		xb_builder_add_locale (builder, locales[i]);
+
+	/* on a read-only filesystem don't care about the cache GUID */
+	if (flags & FU_REMOTES_LOAD_FLAG_READONLY_FS)
+		compile_flags |= XB_BUILDER_COMPILE_FLAG_IGNORE_GUID;
+
+	/* build the metainfo silo */
+	cachedirpkg = fu_common_get_path (FU_PATH_KIND_CACHEDIR_PKG);
+	xmlbfn = g_build_filename (cachedirpkg, "metainfo.xmlb", NULL);
+	xmlb = g_file_new_for_path (xmlbfn);
+	self->silo = xb_builder_ensure (builder, xmlb, compile_flags, NULL, error);
+	if (self->silo == NULL)
+		return FALSE;
+
+	/* load remotes */
+	return fu_remotes_reload (self, error);
+}
+
+GPtrArray *
+fu_remotes_get_all (FuRemotes *self)
+{
+	g_return_val_if_fail (FU_IS_REMOTES (self), NULL);
+	return self->remotes;
+}
+
+FwupdRemote *
+fu_remotes_get_by_id (FuRemotes *self, const gchar *remote_id)
+{
+	g_return_val_if_fail (FU_IS_REMOTES (self), NULL);
+	return fu_remotes_get_by_id_noref (self->remotes, remote_id);
+}
+
+static void
+fu_remotes_class_init (FuRemotesClass *klass)
+{
+	GObjectClass *object_class = G_OBJECT_CLASS (klass);
+	object_class->finalize = fu_remotes_finalize;
+
+	signals[SIGNAL_CHANGED] =
+		g_signal_new ("changed",
+			      G_TYPE_FROM_CLASS (object_class), G_SIGNAL_RUN_LAST,
+			      0, NULL, NULL, g_cclosure_marshal_VOID__VOID,
+			      G_TYPE_NONE, 0);
+}
+
+static void
+fu_remotes_init (FuRemotes *self)
+{
+	self->remotes = g_ptr_array_new_with_free_func ((GDestroyNotify) g_object_unref);
+	self->monitors = g_ptr_array_new_with_free_func ((GDestroyNotify) g_object_unref);
+}
+
+static void
+fu_remotes_finalize (GObject *obj)
+{
+	FuRemotes *self = FU_REMOTES (obj);
+
+	if (self->os_release != NULL)
+		g_hash_table_unref (self->os_release);
+	if (self->silo != NULL)
+		g_object_unref (self->silo);
+	g_ptr_array_unref (self->remotes);
+	g_ptr_array_unref (self->monitors);
```

Comment:
Do the monitors need to be cancelled before unref, or is that implict?

Reply:
They're cancelled on unref.